instr.ml

```diff
@@ -103,59 +103,36 @@ let resolve (code : instruction_stream) (label : string) =
 
 module VarSet = Set.Make(Variable)
 
-type typed_var =
-  | Mut_var of string
-  | Const_var of string
+type variable_type = Mut_var | Const_var
+
+type moded_var = variable_type * variable
 
 exception Incomparable
 
-module TypedVar = struct
-  type t = typed_var
-  let compare a b =
-    match (a,b) with
-    | Mut_var   a, Mut_var   b
-    | Const_var a, Const_var b -> String.compare a b
-    | Mut_var   a, Const_var b
-    | Const_var a, Mut_var   b ->
-      if a = b then raise Incomparable else String.compare a b
+module ModedVar = struct
+  type t = moded_var
+  let compare (ma, a) (mb, b) =
+    match String.compare a b with
+    | 0 ->
+      if ma = mb then 0 else raise Incomparable
+    | c -> c
```

Comment:
I think this is what the code does: `compare` returns `0` when the two inputs are the same, and then we check that their mode are the same otherwise we fail. I tried to follow your (reasonable) behavior on this.